flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionServiceTest.java

```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.leaderelection;
+
+import org.junit.Test;
+
+import java.util.UUID;
+
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link DefaultLeaderElectionService}.
+ */
+public class DefaultLeaderElectionServiceTest {
+
+	private static final String TEST_URL = "akka//user/jobmanager";
+	private static final long timeout = 30L * 1000L;
+
+	@Test
+	public void testOnGrantAndRevokeLeadership() throws Exception {
+		final TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory testingLeaderElectionDriverFactory =
+			new TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory();
+		final DefaultLeaderElectionService leaderElectionService = new DefaultLeaderElectionService(
+			testingLeaderElectionDriverFactory);
+		final TestingContender testingContender = new TestingContender(TEST_URL, leaderElectionService);
+		leaderElectionService.start(testingContender);
+
+		// grant leadership
+		final TestingLeaderElectionDriver testingLeaderElectionDriver =
+			testingLeaderElectionDriverFactory.getCurrentLeaderDriver();
+		assertThat(testingLeaderElectionDriver, is(notNullValue()));
+		testingLeaderElectionDriver.isLeader();
+
+		testingContender.waitForLeader(timeout);
+		assertThat(testingContender.isLeader(), is(true));
+		assertThat(testingContender.getDescription(), is(TEST_URL));
+
+		// Check the external storage
+		assertThat(testingLeaderElectionDriver.getLeaderInformation().getLeaderAddress(), is(TEST_URL));
+
+		// revoke leadership
+		testingLeaderElectionDriver.notLeader();
+		testingContender.waitForRevokeLeader(timeout);
+		assertThat(testingContender.isLeader(), is(false));
+
+		leaderElectionService.stop();
+	}
+
+	@Test
+	public void testLeaderInformationChangedAndShouldBeCorrected() throws Exception {
+		final TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory testingLeaderElectionDriverFactory =
+			new TestingLeaderElectionDriver.TestingLeaderElectionDriverFactory();
+		final DefaultLeaderElectionService leaderElectionService = new DefaultLeaderElectionService(
+			testingLeaderElectionDriverFactory);
+		final TestingContender testingContender = new TestingContender(TEST_URL, leaderElectionService);
+		leaderElectionService.start(testingContender);
+
+		final TestingLeaderElectionDriver testingLeaderElectionDriver =
+			testingLeaderElectionDriverFactory.getCurrentLeaderDriver();
+		assertThat(testingLeaderElectionDriver, is(notNullValue()));
+		testingLeaderElectionDriver.isLeader();
+		testingContender.waitForLeader(timeout);
+
+		// Leader information changed and should be corrected
+		testingLeaderElectionDriver.leaderInformationChanged(LeaderInformation.empty());
+		assertThat(testingLeaderElectionDriver.getLeaderInformation().getLeaderAddress(), is(TEST_URL));
```

Comment:
Shouldn't we also check that the leader session id stays the same?

Reply:
Will check it.